### ISSUE TITLE
Clipboard.hasImageAsync Always Returns False on iOS

### ISSUE THREAD

**Issue by nickcherry** [expo/expo#18114]
### Summary

Below is a demo of the behavior.

https://user-images.githubusercontent.com/229477/177403035-7be006b7-bd54-4791-94ec-315111a882d2.mov

Observe that after copying an image to the clipboard, focusing the app and pressing the "Does clipboard have an image?" button – which checks whether an image is present via `Clipboard.hasImageAsync()` – the app tells us there is no image on the clipboard. However, we can paste the image into other apps.

It's possible this is related to [stricter copy/paste permission imposed by Apple](https://gamingsym.in/ios-16-asks-user-for-permission-to-copy-and-paste-between-apps/), but I have never seen a prompt asking for permission. The functionality works as expected on Android.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

45.0.6

### Environment

  expo-env-info 1.0.4 environment info:
    System:
      OS: macOS 12.4
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.13.1 - ~/.nvm/versions/node/v16.13.1/bin/node
      Yarn: 1.22.17 - ~/.nvm/versions/node/v16.13.1/bin/yarn
      npm: 8.1.2 - ~/.nvm/versions/node/v16.13.1/bin/npm
      Watchman: 2022.03.21.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/nickcherry/.rvm/gems/ruby-3.0.2/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 2021.2 AI-212.5712.43.2112.8609683
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~45.0.0 => 45.0.6
      react: 17.0.2 => 17.0.2
      react-dom: 17.0.2 => 17.0.2
      react-native: 0.68.2 => 0.68.2
      react-native-web: 0.17.7 => 0.17.7
    Expo Workflow: managed

### Reproducible demo

https://github.com/nickcherry/expo-clipboard-paste-image-demo

As per the README, install dependencies by running `yarn install`, then start the iOS app by running `yarn ios`. Copy an image to the clipboard from another app, then focus the demo app and press the "Does clipboard have an image?" button. Observe that `Clipboard.hasImageAsync()` returns `Promise<false>`, even though pasting images works fine in other apps.

**Comment by barthap** [comment]
Interesting... I cannot reproduce it on real iPhone Xs, iOS 15.5

https://user-images.githubusercontent.com/278340/177412242-d43f5ed8-8259-489a-a673-84814c7a98c0.mov

> It's possible this is related to [stricter copy/paste permission imposed by Apple](https://gamingsym.in/ios-16-asks-user-for-permission-to-copy-and-paste-between-apps/), but I have never seen a prompt asking for permission.

It starts with iOS 16. On iOS 15 it should work without any permissions.

What's more interesting, the `hasImageAsync()` directly delegates to the iOS system clipboard API, there's nothing on the expo side in between.

One thing that comes to my mind is that maybe the HTML is copied with the `<img>` tag in it - try checking if clipboard has HTML.

**Comment by nickcherry** [comment]
Ah, I think you are right RE the permissions being an iOS 16-specific problem. I've [updated](https://github.com/nickcherry/expo-clipboard-paste-image-demo/commit/bac42ed3193cddceeb821576e8fb8baf1b4cbba4) the [demo app](https://github.com/nickcherry/expo-clipboard-paste-image-demo) to show a more complete view of the clipboard state. It looks like the image – which was originally a photo taken directly from my camera roll – is being copied as a path string (e.g. `file:///var/mobile/Library/etc`), but still no dice with `hasImageAsync` or `getImageAsync`. This is the behavior on both the simulator and my physical device (iPhone 12 Pro with iOS 15.5).

https://user-images.githubusercontent.com/229477/177453800-eff9f79b-714d-4d5b-902a-a46b06228adf.mov

**Comment by barthap** [comment]
This is the behaviour of the iOS Messages app - it doesn't copy the image, but the file URL to it instead. Expo can do nothing about it. When you copy it from the Gallery app - it works, because the gallery copies the image data, not the URL.

I observed the same thing on Samsung S7 (Android 8 Oreo) - when copying an image from the Notes app - it copied it as part of HTML content and I'd need to extract the `<img>` tag to open it.

The workaround is to check string content if it contains a URL - another weird thing I observed here is that both `getStringAsync` and `getUrlAsync` return that URL, but only `hasStringAsync` is true. Looks like an iOS bug.
